Jobs tests

sofyalaski opened the pull request:
## Description

Implements API tests for jobs authorization endpoints and proposes fixes for missing cases 

## Motivation
Fixes are proposed to follow the authorization logic described in the documentation

## Fixes:
### jobs.controller.ts
* use `CreateJobAuth` enum instead of `AuthOp` to define `jobDatasetAuthorization`
* in  `instanceAuthorizationJobCreate` shift an "if-block" checking the presence of `ownerGroup` down after the admin check was passed (reason: admin should be able to create a job for anonym)
*  in the same function, at the very beginning when initializing values of the JobClass instance add jobParams and leave ownerUser and ownerGroup undefined (for anonymous case)* 
* in the same function, check if the current authorization is in `jobDatasetAuthorization` (types  mismatching) 
*  in the same function,  add a check if an unauthorized user creates a job for some other group (unauthorized users should only create jobs for themselves)
* in the same function, for "dataset"-kind of job configurations, the "where" filtering wasn't grabbing the respective dataset values properly (maybe go back to Despina's implementation)
* in the `checkDatasetIds` function add a check if each requested datasetId exists in database ( for the "dataset" -kind of job configurations)
* implement `generateJobInstanceForPermissions` function, which is similar to the one in datasets.controller. mongoose `findOne` and `findAll` do not return an instance of JobClass as implemented but a simple JS Object. Since we still need to pass that object to casl( which expects the instance of JobClass) we need conversion. To not write all properties of JobClass, a minimal copy with only properties required for permission checking is created. The return instances are not affected by this and remain  JS Object.
*  functionality of `instanceAuthorizationJobStatusUpdate` is now implemented in the patch endpoint function. This is done to avoid redefining a job return by `findOne` as a instance of JobClass with all properties.
* for both GET endpoints only the endpoint authorisation was implemented. To add the instance authorization too, first create JobClass instance of the found job Objects (as described in two above items) and pass them to respective casl expressions. 
* adds implementation of delete endpoints. The rules are defined in the CheckPolicies decorator 

### jobs.service.ts 
* add a function to extract the username of a user sending a request ( should return "anonymous" if request sent by unauthenticated user). This is used both in `create` and`statusUpdate` functions 

### casl-ability.factory.ts   
* for statusUpdate endpoint the user from update_job_groups had less rights that a normal user -> fixed that 
* add instance authorization for statusUpdate for anonymous user
* implement delete endpoint authorization  
* fixed a negation typo 

## Changes
* add deleteJobGroups to `configuration.ts`

## Tests included

- [ ] Included for each change/fix?
- [ ] Passing? (Merge will not be approved unless this is checked)

Junjiequan committed:
feat: reset dataset before every api test (#1131)

* reset dataset before every api test

* Clear collections before each api test

sofyalaski committed:
first tests

sofyalaski committed:
Fixes after jobs testing:
1. Use CreateJobAuth enum instead of AuthOp to define jobDatasetAuthorization
2. Shift a block if groupOwner is present down after the admin check was passed (admin create a job for anonym)
3. Check if current authorization is in jobDatasetAuthorization in instanceAuthorizationJobCreate (types  mismatching)
4. Add a check if unauthorized user creates a job for someone else

sofyalaski committed:
Fixes after jobs testing:
if anonymous user creates a job, initialize a jwt instance for him/her

sofyalaski committed:
Fixes after testing:
string array includes string enum

sofyalaski committed:
dataset(s)Validation redundancy after a merged PR

sofyalaski committed:
test Jobs authorization wip

sofyalaski committed:
fix filtering syntax

sofyalaski committed:
fixes after Job testing:
1. comment out jobConfig matchin, as now in interceptor
2. fix filtering object hierarchy
in #dataset kind of jobConfig's
3. not defining jobInstance ownerUser/Group as empty string
4. add jobInstance.jobParams initialization
5. add a check if datasets passed in the ID list exist

sofyalaski committed:
minor changes

sofyalaski committed:
fix findOne filtering syntax

sofyalaski committed:
add Jobs tests

sofyalaski committed:
remove redundancy

sofyalaski committed:
fix numbering

sofyalaski committed:
extend UPDATE_JOB_GROUP user's right to match docs

sofyalaski committed:
remove unneccessary arguments

sofyalaski committed:
make function return value of findOne  an instance of JobClass

sofyalaski committed:
add statusUpdate tests

sofyalaski committed:
add jobParams to the job instance

sofyalaski committed:
Patch tests

sofyalaski committed:
merge changes from release-jobs

sofyalaski committed:
Fixes after tests:
* Anonymous user can patch status update to a job in #all config. Added instance authorization

* Implement JobDelete authorization

sofyalaski committed:
add deleteJobGroups from env to configuration

sofyalaski committed:
add username for anonymous user when creating/updating status of the job

sofyalaski committed:
statusUpdate instead of update

sofyalaski committed:
Fixes after testing:

* similarly to Datasets, mongoose returns an Object and not JobClass instances. To pass this into CASL we first need  to make those class instances. generateJobInstanceForPermissions implements a copy of object as correct class instance only with properties required in casl.

* functionality of instanceAuthorizationJobStatusUpdate is now implemented in the patch endpoint function. This is done to avoid redefining a found job as a instance of JobClass with all properties

* For both GET endpoints only the endpoint authorisation was implemented for user in the functions of the controller. To add the instance authorization too, first create JobClass instance of the found job Objects and pass them to respective casl expressions.

* adds implementation of delete endpoints. The rules are defined in the CheckPolicies decorator.

sofyalaski committed:
Finalized tests for authorization

sofyalaski committed:
add delete_job_groups definition

sofyalaski committed:
fix unit tests after test jobconfig.json changes

despadam commented on test/config/jobconfig.json:
```diff
@@ -1,53 +1,80 @@
   {
-    "jobType": "archive",
+    "jobType": "all_access",
+    "configVersion": "v1.0 2024-03-01 6f3f38",
```
There is a `configVersion` per job but also a general one at the top of the file. We should only keep the top one. the current implementation expects only that and then stores this information inside each job, but it's not needed to duplicate it per job in the configuration file.

sofyalaski replied:
Resolved in the latest commit

despadam commented on test/Jobs.js:
```diff
@@ -1,33 +1,151 @@
+
+describe("1100: Jobs: Test New Job Model", () => {
+  before(() => {
+    db.collection("Dataset").deleteMany({});
+    db.collection("Job").deleteMany({});
```
I have been trying to figure out why this part is not working. We expect that it would empty the `Dataset` and `Job` collections, so that we can start the tests with an empty database, however this does not happen. 

One idea to solve this would be to prune the `mongodb volume` before starting the test, but I do not know whether this is the right approach. 

In my effort to understand what these two lines of code are doing, I came across the file `pretest.js` file, which contains the following:
```
const uri = "mongodb://localhost:27017/scicat";
const client = new MongoClient(uri);
global.db = client.db("scicat");
```
So now my question is why are we trying to connect to `scicat` db? Shouldn't it be `test_be_next`? Or perhaps there is something that I don't quite understand. Could someone please this to me? Perhaps @sofyalaski or @Junjiequan ?

Junjiequan replied:
It's my mistake. the db name should be customizable. In your case it didn't work because you are not creating db named `scicat`, so the local test would fail. I believe there's a ENV variable named ~~test_db or sth~~ `MONGODB_URI`, we can change to that instead hardcode here. I will fix it next week and open a PR.
Thanks for mentioning this.

sofyalaski committed:
empty collections before testing

despadam commented on src/jobs/jobs.controller.ts:
```diff
@@ -319,9 +314,57 @@ export class JobsController {
+        throw new HttpException(
+          {
+            status: HttpStatus.BAD_REQUEST,
+            message: "At least one dataset in the ids list doesn't.",
```
The error message here is not clear.

despadam commented on src/jobs/jobs.controller.ts:
```diff
@@ -46,6 +48,9 @@ import {
 import { JobCreateInterceptor } from "./interceptors/job-create.interceptor";
+import { ConnectableObservable } from "rxjs";
+import { integer } from "@elastic/elasticsearch/lib/api/types";
+import { boolean } from "mathjs";
```
Are these needed?

despadam commented on src/jobs/jobs.controller.ts:
```diff
@@ -16,13 +16,15 @@ import {
 } from "@nestjs/common";
 import { Request } from "express";
 import { FilterQuery } from "mongoose";
+import { plainToClass, plainToInstance } from 'class-transformer';
```
Are these needed?

despadam commented on src/jobs/jobs.controller.ts:
```diff
@@ -368,11 +404,20 @@ export class JobsController {
           );
         }
-
         jobInstance.ownerUser = jobUser?.username as string;
         jobInstance.contactEmail = jobUser?.email as string;
         jobInstance.ownerGroup = jobCreateDto.ownerGroup;
```
What is the expected result of these lines? I tested it by logging in as an admin and creating a job without providing ownerUser or ownerGroup. I am printing the `jobInstance` after these lines but everything seems to be undefined: createdBy, ownerGroup,, ownerUser, contactEmail are all undefined.

sofyalaski replied:
"createdBy" field should be created by `utils.addCreatedByFields` from `jobs.service` and when I am logged in, it usually fills that field. Are you sure you were logged in? For the other fields, they are left empty for the case where admin creates a job for anonymous user. Then it's important to specify the email. For other cases, where users exist in the database, the email is going to be grabbed from the user data.

despadam commented on src/jobs/jobs.controller.ts:
```diff
@@ -390,44 +435,76 @@ export class JobsController {
+        }
       // verify that the user meet the requested permissions on the datasets listed
-      const datasetIds = this.checkDatasetIds(jobCreateDto.jobParams);
+      // const datasetIds = await this.checkDatasetIds(jobCreateDto.jobParams);
```
Why is this commented out?

sofyalaski replied:
it's already called in line 392. 
The reasoning behind that was that datasets were only checked when the configuration was set to "#datasetX". Now they are being checked for all configuration if "datasetIds" are passed to jobParams.

sbliven replied:
So delete it?

despadam commented on src/jobs/jobs.controller.ts:
```diff
@@ -572,11 +626,18 @@ export class JobsController {
+
+    const ability = this.caslAbilityFactory.createForUser(request.user as JWTUser);
+    // check if he/she can create this dataset
+    const canCreate =
```
Change comment and variable name to `update` instead of `create`

despadam commented on src/jobs/jobs.controller.ts:
```diff
@@ -610,7 +671,25 @@ export class JobsController {
+    const canCreate =
+      ability.can(AuthOp.JobReadAny, JobClass) ||
+      ability.can(AuthOp.JobReadAccess, currentJobInstance);
+    if (!canCreate) {
+      throw new ForbiddenException("Unauthorized to update this dataset");
```
Rename variable and change error message to be representative to GET

despadam commented on src/jobs/jobs.controller.ts:
```diff
@@ -670,9 +764,7 @@ export class JobsController {
-  @CheckPolicies(
-    (ability: AppAbility) => ability.can(AuthOp.Delete, JobClass), // TBD
-  )
+  @CheckPolicies((ability: AppAbility) => ability.can(AuthOp.JobDelete, JobClass)&& ability.can(AuthOp.JobDeleteAny, JobClass) )
```
Need some more details on this. Do we need to check whether the specific used has the rights to delete a specific job? Is this check enough?

When logged in as an admin, I tried to delete a job that I created but got error `403: forbidden resource`

sofyalaski replied:
Yes, If you look in the [docs](https://scicatproject.github.io/documentation/Development/v4.x/backend/authorization/authorization_jobs.html) (first table) admin shouldn't have rights to delete jobs, unless he/she is in the DELETE_JOB_GROUPS as defined in env file. I put archiveManager similar to datasets.

despadam reviewed (changes requested):
Reviewed `jobs.controller`

sofyalaski committed:
changes requested in PR

sofyalaski commented:
@despadam I made some changes you requested. For other comments, we can discuss tomorrow in person

sofyalaski committed:
eremove version duplicate

sofyalaski committed:
Merge remote-tracking branch 'upstream/release-jobs' into jobs-tests
merge chanhes after conversation on how configuration will be stored

sofyalaski committed:
change getJobMatchingConfiguration function to explicitly use job type, st it can be used for all cases where we need to extract the configuration, without having to pass the full dto

sofyalaski committed:
run full testing workflow on pushing to release branches

sofyalaski commented:
@Despina, I've updated the change to configuration grabbing for statusUpdate as  we discussed

sofyalaski committed:
add env variable for jobs config path for github workflow

sbliven committed:
Fix lint errors

- Run lint:fix
- Ignore no-explicit-any rule for jobParams (which are not type checked)
- remove 'read' jobOperation (not implemented)

Junjiequan commented on test/config/pretest.js:
```diff
@@ -3,7 +3,7 @@
 var chaiHttp = require("chai-http");
 const { MongoClient } = require("mongodb");
-const uri = "mongodb://localhost:27017/scicat";
+const uri = process.env.MONGODB_URI;
```
If we set the test DB URI as the MONGODB_URI, it will wipe the local development DB if someone forgets to change the URI before running the test. Therefore, I suggest keeping it unchanged for now.

sbliven replied:
Good point. `appUrl` is also hard-coded a few lines down. Unless there was a particular reason this needs to be changed at runtime?

Junjiequan replied:
I don't think we need `appUrl` to be changed at run time. I'm not sure if we have environment variables to assign to the `appUrl`

Junjiequan replied:
After discussing with @nitrosx regarding the test database, he suggested using the MONGODB_URI variable for the test database as well. In the documentation, we can mention that one should change the MONGODB_URI before running the test command to prevent accidental local data wipes.

Here are some minor suggestions for the code changes:

- Include require("dotenv").config() to load the environment variables. Without this, the environment variables will be undefined.
- For global.db, replace client.db(uri.split("/").pop()); with client.db().

```suggestion
require("dotenv").config();
var chaiHttp = require("chai-http");

const { MongoClient } = require("mongodb");

const client = new MongoClient(process.env.MONGODB_URI);

async function loadChai() {
  const { chai } = import("chai");
  chai.use(chaiHttp);
  await client.connect();
}

loadChai();
global.appUrl = "http://localhost:3000";
global.request = require("supertest");
global.db = client.db();
```

sbliven committed:
Merge branch 'release-jobs' into jobs-tests

sbliven committed:
Fix lint following merge

sbliven commented on src/jobs/config/jobconfig.ts:
```diff
@@ -54,25 +50,21 @@ export class JobConfig {
-    jobData: Record<string, any>,
-    configVersion: string
-  ): JobConfig {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
```
Is `Record<string, any>` the best type for jobData? I had to disable linting because of the `any` here and in all subclasses. I think that's reasonable though, since it really is unchecked.

sbliven replied:
I realized that `unknown` fits better here than any.

sbliven commented on src/jobs/actions/urlaction.ts:
```diff
@@ -32,12 +27,14 @@ export class URLAction<T> implements JobAction<T> {
     return URLAction.actionType;
   }
 
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Weird that this empty method causes a lint warning; it overrides the interface method.

sbliven commented on src/jobs/jobs.controller.ts:
```diff
@@ -320,31 +310,77 @@ export class JobsController {
+          pid: { $eq: datasetIds[i] },
+        },
+      };
+      const idCount = await this.datasetsService.count(filter);
```
Wouldn't it be more performant to make a single database request here, rather than one request per datasetId?

sofyalaski replied:
Yes, you're right, I changed that in the last commit

sbliven commented on src/jobs/jobs.controller.ts:
```diff
@@ -320,31 +310,77 @@ export class JobsController {
+        throw new HttpException(
+          {
+            status: HttpStatus.BAD_REQUEST,
+            message: "At least one dataset in the ids list doesn't exist.",
```
Should include `datasetIds[i]` in the error message to make debugging easier.

sofyalaski replied:
addressed in the last commit

sbliven commented on src/jobs/jobs.controller.ts:
```diff
@@ -758,6 +866,7 @@ export class JobsController {
     @Req() request: Request,
     @Param("id") id: string,
   ): Promise<unknown> {
+    Logger.log("Deleting job!");
```
Should include the id in the log message

sofyalaski replied:
addressed in the last commit

sbliven commented on test/Jobs.js:
```diff
@@ -40,755 +158,3707 @@ describe.skip("1100: Jobs: Test New Job Model", () => {
+                                password: TestData.Accounts["archiveManager"]["password"],
+                              },
+                              (tokenVal) => {
+                                accessTokenArchiveManager = tokenVal;
```
That's a lot of nesting! Not necessary for this PR, but maybe `getToken` needs to be refactored to return a promise rather than use callbacks.

sofyalaski replied:
This one is addressed in PR #1350

sbliven commented on test/Jobs.js:
```diff
@@ -40,755 +158,3707 @@ describe.skip("1100: Jobs: Test New Job Model", () => {
+  //       res.body.should.not.have.property("id");
+  //       res.body.should.have.propert("message").and.be.equal("Dataset ids list was not provided in jobParams");
+  //     });
+  // });
```
Why commented?

sofyalaski replied:
yes, I was confused by the logic myself a bit. I changed this one into 0065. Now these tests at the beginning cover:
- `jobParams` is passed with property `datasetIds`, but the array is empty --> fail (0040)
- `jobParams` is passed with property `datasetIds`, but it refers to datasetIds that are not in database --> fail (0050)
- `jobParams` is not passed at all --> fail (0060)
- `jobParams` is passed as empty object --> fail (0065)

I additionally implemented the check for `jobParams` property in the `create` endpoint. It checks that `jobParams` is passed and is not empty. Before that if no `jobParams` were passed a 500 internal error was thrown. Now the error message is explicitly saying that `jobParams` need to be defined and not empty.   

the last two tests are new. 0060 before was testing with an empty object as jobParams, which passed before. Now, as I I changed the implementation a little, this test fails. I think it is logical that if a job has no context it shouldn't be accepted? This test now is numbered as 0065.
0060 tests the case where no `jobParams` passed at all.

despadam committed:
Update test/config/pretest.js to use "dotenv"

Co-authored-by: Jay <a331998513@gmail.com>

despadam committed:
Fix pretest.js after merging suggestions

despadam committed:
Fix pretest.js

sbliven committed:
Fix jobconfig no-explicit-any issues

- Revert previous changes to ignore no-explicit-any
- Use `unknown` for jobParam inputs plus stronger runtime checks
- Fix compilation error in elastic-search service relating to
  deleting non-optional variables. 
- Remove rabbitmqaction validation. RabbitMQ should ignore the DTO body,
  and validate the config itself in the constructor.

sbliven committed:
Fix jobconfig no-explicit-any issue (missed logaction)

sbliven reviewed (approved):
This looks pretty good. I added a couple minor comments but it looks good.

Can you verify that mocha passes before merging? @despadam and I both ran into issues running the tests. Mocha should be added to the CI following this PR.

despadam commented on src/jobs/actions/rabbitmqaction.ts:
```diff
@@ -12,80 +11,99 @@ export class RabbitMQJobAction<T> implements JobAction<T> {
-    }
-  }
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  async validate(dto: T) {}
```
@sbliven I struggle to understand the need for `validate()` in the `action class`. I was always wondering whether it makes sense validating after calling the constructor, and your changes here I think prove my point that it probably doesn't. I would suggest following this example and moving all validation inside the constructor of each action. Then:
- On job `create` if creating an action fails, the job creation should also fail.
- On job `statusUpdate` things get a bit more complicated. Since we are using a global job configuration, instead of job specific, if no changes were made to the configuration then no validation is needed at all and we can proceed directly to performing the job. But, if the configuration has changed, we would need a method that would first update the values of the class' attributes, before performing the action.

despadam replied:
After discussing with @sofyalaski, I will create a separate PR to address this.

despadam replied:
Will be covered by validate action

sofyalaski committed:
changes in tests:
0060 now checks if a job was passed without a required parameter JobParams
0065 (new) checks if a job was passed with an empty object for jobParams
1950 now tries to delete a job that doesn't exist and fails
1960 was called 1950 before
counts on GET are changed (because 0060 before was passing)

sofyalaski committed:
address comments in the PR:
*  now makes one request to the db to find datasets and extracts their pids.
*  loggs the names of all dataset pids that are not in the db in case of an error
*  job endpoint implements the check that the jobParams were passed and are not empty, otherwise it will throw a bad request with respective message
*  job endpoint loggs the id of job to be deleted and implements a check if job exist in db, otherwise throws an error.

sofyalaski commented:
@sbliven in[ the commit](https://github.com/SciCatProject/scicat-backend-next/pull/1286/commits/9fa1b8ef069f9a924e2493abc70a42400e78dfbb) two weeks ago I added running the tests on PR to the "relase-*" branches, and you can see the bars here at the bottom, I assume. For me, the API tests seem to be passing (API test include both `test:api:jest` and `test:api:mocha`. It might be because of some differences in the way we deploy then? We tried running the brand new docker container and running tests in it with Despina and it worked. Also, sorry for not telling that you need to start the app first!
